amm/contracts/stable_pool/amp_coef.rs

```diff
@@ -0,0 +1,231 @@
+    ensure,
+};
+use ink::env::DefaultEnvironment;
+use traits::{MathError, StablePoolError};
+
+#[derive(Default, Debug, scale::Encode, scale::Decode, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+)]
+pub struct AmpCoef {
+    /// Initial amplification coefficient.
+    init_amp_coef: u128,
+    /// Target for ramping up amplification coefficient.
+    target_amp_coef: u128,
+    /// Initial amplification time.
+    init_amp_time: u64,
+    /// Stop ramp up amplification time.
+    stop_amp_time: u64,
+}
+
+impl AmpCoef {
+    pub fn new(init_amp_coef: u128) -> Result<Self, StablePoolError> {
+        ensure!(init_amp_coef >= MIN_AMP, StablePoolError::AmpCoefTooLow);
+        ensure!(init_amp_coef <= MAX_AMP, StablePoolError::AmpCoefTooHigh);
+        Ok(Self {
+            init_amp_coef,
+            target_amp_coef: init_amp_coef,
+            init_amp_time: 0,
+            stop_amp_time: 0,
+        })
+    }
+
+    pub fn compute_amp_coef(&self) -> Result<u128, MathError> {
+        let current_time = ink::env::block_timestamp::<DefaultEnvironment>();
+        if current_time < self.stop_amp_time {
+            let time_range = self
+                .stop_amp_time
+                .checked_sub(self.init_amp_time)
+                .ok_or(MathError::SubUnderflow(51))?;
+            let time_delta = current_time
+                .checked_sub(self.init_amp_time)
+                .ok_or(MathError::SubUnderflow(52))?;
+
+            // Compute amp factor based on ramp time
+            let amp_range = self.target_amp_coef.abs_diff(self.init_amp_coef);
+            let amp_delta = amp_range
+                .checked_mul(time_delta as u128)
+                .ok_or(MathError::MulOverflow(51))?
+                .checked_div(time_range as u128)
+                .ok_or(MathError::DivByZero(51))?;
+            if self.target_amp_coef >= self.init_amp_coef {
+                // Ramp up
+                self.init_amp_coef
+                    .checked_add(amp_delta)
+                    .ok_or(MathError::AddOverflow(1))
+            } else {
+                // Ramp down
+                self.init_amp_coef
+                    .checked_sub(amp_delta)
+                    .ok_or(MathError::SubUnderflow(55))
+            }
+        } else {
+            Ok(self.target_amp_coef)
+        }
+    }
+
+    pub fn ramp_amp_coef(
+        &mut self,
+        future_amp_coef: u128,
+        future_time_ts: u64,
+    ) -> Result<(), StablePoolError> {
+        ensure!(future_amp_coef >= MIN_AMP, StablePoolError::AmpCoefTooLow);
+        ensure!(future_amp_coef <= MAX_AMP, StablePoolError::AmpCoefTooHigh);
+        let current_time = ink::env::block_timestamp::<DefaultEnvironment>();
+        let ramp_duration = future_time_ts.checked_sub(current_time);
+        ensure!(
+            ramp_duration.is_some() && ramp_duration.unwrap() >= MIN_RAMP_DURATION,
```

Comment:
This should be OK due to lazy evaluation of `&&`. If `ramp_duration` is `None` then `unwrap` won't happen.
Although it would be safer to have a unit test case which tries `future_time_ts` lower than `current_time` and checks for this particular `AmpCoefRampDurationTooShort` error.